events/suite_test.go

```diff
@@ -0,0 +1,179 @@
+	RegisterFailHandler(Fail)
+	RunSpecs(t, "Events")
+}
+
+var _ = BeforeSuite(func() {
+	fakeClock = clock.NewFakeClock(time.Now())
+	kubeClient = fake.NewClientBuilder().WithScheme(scheme.Scheme).WithIndex(&corev1.Event{}, "involvedObject.kind", func(o client.Object) []string {
+		evt := o.(*corev1.Event)
+		return []string{evt.InvolvedObject.Kind}
+	}).Build()
+	controller = events.NewController[*TestObject](kubeClient, fakeClock)
+	ctx = log.IntoContext(context.Background(), ginkgo.GinkgoLogr)
+})
+
+var (
+	SchemeBuilder = runtime.NewSchemeBuilder(func(scheme *runtime.Scheme) error {
+		scheme.AddKnownTypes(schema.GroupVersion{Group: test.APIGroup, Version: "v1alpha1"}, &TestObject{})
+		return nil
+	})
+)
+
+var _ = Describe("Controller", func() {
+	BeforeEach(func() {
+		controller.Reset()
+	})
+	It("should emit metrics on an event", func() {
+		events := []*corev1.Event{}
+		for i := range 5 {
+			events = append(events, createEvent("test-object", fmt.Sprintf("Test-type-%d", i), fmt.Sprintf("Test-reason-%d", i)))
+			ExpectApplied(ctx, kubeClient, events[i])
+		}
+
+		for i := range 5 {
+			Expect(getMetric("operator_testobject_event_total", conditionLabels(fmt.Sprintf("Test-type-%d", i), fmt.Sprintf("Test-reason-%d", i)))).To(BeNil())
+		}
+
+		for i := range 5 {
+			_, err := reconcile.AsReconciler(kubeClient, controller).Reconcile(ctx, reconcile.Request{NamespacedName: client.ObjectKeyFromObject(events[i])})
+			Expect(err).ToNot(HaveOccurred())
+		}
+
+		for i := range 5 {
+			Expect(getMetric("operator_testobject_event_total", conditionLabels(fmt.Sprintf("Test-type-%d", i), fmt.Sprintf("Test-reason-%d", i))).GetCounter().GetValue()).To(BeEquivalentTo(1))
+		}
+	})
+	It("should not fire metrics if the last transition was before controller start-up", func() {
+		event := createEvent("test-name", corev1.EventTypeNormal, "reason")
+		event.LastTimestamp.Time = time.Now().Add(30 * time.Minute)
+		ExpectApplied(ctx, kubeClient, event)
+
+		Expect(getMetric("operator_testobject_event_total", conditionLabels(corev1.EventTypeNormal, "reason"))).To(BeNil())
+
+		_, err := reconcile.AsReconciler(kubeClient, controller).Reconcile(ctx, reconcile.Request{NamespacedName: client.ObjectKeyFromObject(event)})
+		Expect(err).ToNot(HaveOccurred())
+
+		Expect(getMetric("operator_testobject_event_total", conditionLabels(corev1.EventTypeNormal, "reason")).GetCounter().GetValue()).To(BeEquivalentTo(1))
+
+		event.LastTimestamp.Time = time.Now().Add(-30 * time.Minute)
+		ExpectApplied(ctx, kubeClient, event)
+		_, err = reconcile.AsReconciler(kubeClient, controller).Reconcile(ctx, reconcile.Request{NamespacedName: client.ObjectKeyFromObject(event)})
+		Expect(err).ToNot(HaveOccurred())
+
+		Expect(getMetric("operator_testobject_event_total", conditionLabels(corev1.EventTypeNormal, "reason")).GetCounter().GetValue()).To(BeEquivalentTo(1))
+	})
+})
+
+// +k8s:deepcopy-gen=true
+// +kubebuilder:object:root=true
```

Comment:
Why do we need to copy over the DeepCopies? Don't we have a package for our test objects?